### ISSUE TITLE
First-party isolation breaks Toolbox

### ISSUE THREAD

**Issue by Geo1088** [toolbox-team/reddit-moderator-toolbox#130]
Our `chrome.cookie` calls don't support first-party isolation. The way we handle this in the background page needs to be updated.

Documentation: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies#First-party_isolation

Reddit thread: https://www.reddit.com/r/toolbox/comments/cmzpfg/bugtoolbox_not_loading_on_6801_firefox/

**Comment by creesch** [comment]
I have been thinking about how we could implement this and I assume we simply no longer directly rely on the session being there but want to send the cookie with the requests. 

If that works we also might have an inroad into our firefox incognito issue (though that is going to be much more complex). 

Looking at all the documentation we might be able to do something along the lines of 

```javascript
    // Only do this stuff with firefox as chrome doesn't know the firstPartyDomain option.
    if(browser) {
        // fetch all cookies for reddit.
        const cookies = await browser.cookies.getAll({
            domain: '.reddit.com',
            firstPartyDomain: 'reddit.com',
        });

        // make the cookies into a header string. Don't be picky, just send wat is stored for reddit.com
        const cookieArray = [];

        cookies.forEach(cookie => {
            cookieArray.push(`${cookie.name}=${cookie.value}`)
        });
        
        const cookieString = cookieArray.join('; ');
        console.log(cookieString)
        
        // Make sure we can set credentials.
        options.xhrFields = {
            withCredentials: true
        }
        // Set beforeSend to add the header
        options.beforeSend = jqXHR => jqXHR.setRequestHeader('Cookie', cookieString);
        
    }

    // And make the request
    makeRequest(options, sendResponse);
```

at this point https://github.com/toolbox-team/reddit-moderator-toolbox/blob/master/extension/data/background/background.js#L438

I didn't test it with the `firstPartyDomain` option yet as it is unclear to me how it works exactly as with it not enabled it only works when it is empty. So there might be some challenges along that point.

**Comment by creesch** [comment]
Well fuck... 

![image](https://user-images.githubusercontent.com/2502190/63009309-88b47880-be84-11e9-9fdd-528c3f8a74a6.png)

[links to this](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSMissingAllowOrigin?utm_source=devtools&utm_medium=firefox-cors-errors&utm_campaign=default) even though it appears reddit does have the header set.

I wonder if it is an xhr limitation and would work with fetch but I am not very hopeful. 

Frankly it appears that they once again forgot to properly include webextensions in the development of this feature making it next to impossible to deal with properly. Because to be 100% honest this shouldn't affect an extension that has all the necessary permissions in the manifest.

**Comment by Geo1088** [comment]
We can also just catch an error from `cookie.get` and retry passing `{firstPartyDomain: 'reddit.com'}` I'm pretty sure... It shouldn't need to be too complicated. There's a method to check if first-party isolation is enabled with the `chrome.privacy` API, but we don't want to do that since it requires more permissions, so I'm gonna try doing it with the catch method and see if I can get it working that way.

**Comment by creesch** [comment]
That isn't the problem though, it appears we are simply not allowed to set cookies due to CORS being weird in this instance. I also tried it with fetch and get redirected to login.reddit.com so something isn't quite right there either.

**Comment by Geo1088** [comment]
@PrinceKael This should be fixed for the next release! I just pushed cfd09e2 which fixed the issue for me in FF 72.0.2.
